invenio_app_rdm/stats/event_builders.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 TU Wien.
+#
+# Invenio App RDM is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Custom event builders for the InvenioRDM statistics.
+
+Note that the arguments to these functions will be the same as passed to the
+``EventEmitter`` objects when they are called.
+Currently in InvenioRDM this is done in resources (for API) and view functions (for UI).
+As such, it is assumed that a request context is available (and thus, Flask's global
+``request`` is accessible).
+"""
+
+from datetime import datetime
+
+from flask import request
+from invenio_stats.utils import get_user
+
+
+def record_view_event_builder(event, sender_app, **kwargs):
+    """Build a record-view event.
+
+    *Note* that this function assumes a request context by accessing properties of
+    Flask's global ``request`` object.
+    """
+    assert "record" in kwargs
+    record = kwargs["record"]
+    event.update(
+        {
+            # When:
+            "timestamp": datetime.utcnow().isoformat(),
+            # From where:
+            "referrer": request.referrer,
+            # Who:
+            **get_user(),
+            # TODO probably we can add more request context information here for
+            #      extra filtering (e.g. discarding the event when it's a
+            #      citation text export)
+        }
+    )
+    return event
+
+
+def check_if_via_api(event, sender_app, **kwargs):
+    """Check if the event comes from an API request.
+
+    *Note* that this function assumes a request context by accessing properties of
+    Flask's global ``request`` object.
+    """
+    via_api = None
+    if "via_api" in kwargs:
+        via_api = kwargs["via_api"]
+    else:
+        # TODO: this fallback heuristic is very opinionated and should be reworked
+        #       maybe we should check if the current_app is the API app?
+        via_api = "api" in request.path
+
+    event.update({"via_api": via_api})
+    return event
+
+
+def drop_if_via_api(event, sender_app, **kwargs):
+    """Drop the event if it comes from the API."""
+    if "via_api" not in event or not event["via_api"]:
+        return event
+
+    return None
+
+
+def add_record_metadata(event, sender_app, **kwargs):
+    """Add metadata of the record in question."""
+
+    # NOTE: the selection of fields is based on the Zenodo stats module
+    #       (see: zenodo.modules.stats.utils:extract_event_record_metadata)
+    assert "record" in kwargs
+    record = kwargs["record"]
+    event.update(
+        {
+            "recid": record.get("id", None),
+            "parent_recid": record.parent.get("id", None),
```

Comment:
should we add information about the records' communities here, similar to what zenodo stores?
how about other fields, like owners, etc.?

also, this could probably be merged together with the `record_view_event_builder`... i don't think anybody would want either of these functions separately :eyes:

Reply:
Merged the builders, created a follow-up issue: https://github.com/inveniosoftware/invenio-app-rdm/issues/2091